gbcamextract.c

```diff
@@ -1,352 +1,245 @@
-    // draw the tile
-    for( spanNum = 0; spanNum<8; spanNum++ )
-    {
-      spanAddress = tileAddress + 2*spanNum;
-      lowBits = frameBuffer[spanAddress];
-      highBits = frameBuffer[spanAddress+1];
-      x = 144;
-      y = 16 + yTile*8 + spanNum;
-      drawSpan( pixelBuffer, lowBits, highBits, x, y );
-    }
-    // rightmost tile
-    tileNum = frameBuffer[frameAddress + 0x650 + yTile*4 + 3];
-    tileAddress = frameAddress + tileNum*16;
-    // draw the tile
-    for( spanNum = 0; spanNum<8; spanNum++ )
-    {
-      spanAddress = tileAddress + 2*spanNum;
-      lowBits = frameBuffer[spanAddress];
-      highBits = frameBuffer[spanAddress+1];
-      x = 152;
-      y = 16 + yTile*8 + spanNum;
-      drawSpan( pixelBuffer, lowBits, highBits, x, y );
-    }
+    tile = frameBuffer + tileAddress;
+    x = xTile*8;
+    y = ((z&1)?8:0) + ((z&2)?128:0);
+    drawSpan( pixelBuffer, tile, x, y );
   }
 }
 
-void drawSpan( char pixelBuffer[static WIDTH*HEIGHT], char lowBits, char highBits, int x, int y )
-{
-  int pixelNum;
-  char spanColors[8];
-  int grays[4] = {0xFF, 0xAA, 0x55, 0x00};
-  for(pixelNum=0; pixelNum<8; pixelNum++)
-  {
-    int color = 0;
-    int mask = 1<<pixelNum;
-    if( lowBits & mask )
-      color = 1;
-    if( highBits & mask )
-      color += 2;
-    spanColors[7-pixelNum] = grays[ color ];
-  }
-  for(pixelNum=0; pixelNum<8; pixelNum++)
+void drawSpan( char pixelBuffer[], char *buffer, int x, int y ) {
+  unsigned char lowBits, highBits, i;
+  char *p;
+  p = pixelBuffer + (x/4) + 1 + y * ROW_SIZE;
+  for( i = 8; i; --i, p+=ROW_SIZE+2 )
   {
-    // copy span to pixel buffer
-    int pixelAddress = x + 160 * y +pixelNum;
-    pixelBuffer[pixelAddress] = spanColors[pixelNum];
+    lowBits = ~*buffer++;
+    highBits = ~*buffer++;
+    *p-- = (lowBits & 1) | ((highBits & 1) << 1) | ((lowBits & 2) << 1) | ((highBits & 2) << 2) | ((lowBits & 4) << 2)  | ((highBits & 4) << 3) | ((lowBits & 8) << 3) | ((highBits & 8) << 4);
+    lowBits >>= 4;
+    highBits >>= 4;
+    *p-- = (lowBits & 1) | ((highBits & 1) << 1) | ((lowBits & 2) << 1) | ((highBits & 2) << 2) | ((lowBits & 4) << 2)  | ((highBits & 4) << 3) | ((lowBits & 8) << 3) | ((highBits & 8) << 4);
   }
 }
 
-void writeImageFile( char pixelBuffer[static WIDTH*HEIGHT], int picNum )
+void writeImageFile( char pixelBuffer[], int picNum )
 {
+  int y;
   // open file
-  char name[50];
+  char name[7];
   sprintf(name, "%d.png", picNum );
   FILE *fp = fopen( name, "wb" );
-  
+
   png_structp png_ptr = png_create_write_struct
     (PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
 
   if (!png_ptr)
-      return;
+      exit(EXIT_FAILURE);
 
   png_infop info_ptr = png_create_info_struct(png_ptr);
   if (!info_ptr)
   {
       png_destroy_write_struct(&png_ptr,
           (png_infopp)NULL);
-      return;
+      exit(EXIT_FAILURE);
   }
-  
+
     // init output
   png_init_io( png_ptr, fp );
-  
+  png_set_compression_level(png_ptr, Z_BEST_COMPRESSION);
```

Comment:
These changes seemed to make the program a bit faster, so I bumped the compression level up to maximum. It seems to save a hundred bytes or so per image, and is not perceptibly slower.